src/databroker_elasticsearch/elasticindex.py

```diff
@@ -0,0 +1,216 @@
+        The default is "run_start".
+    doc_properties : dict
+        The field names and data types in the Elasticsearch `doc_type`. [1]_
+
+    References
+    ----------
+    .. [1] https://www.elastic.co/guide/en/elasticsearch/guide/current/mapping.html
+    """
+
+    def __init__(
+            self,
+            es,
+            index: str,
+            mapper: Callable=None,
+            criteria: Callable=None,
+    ):
+        self.es = (Elasticsearch(es) if isinstance(es, str)
+                   else Elasticsearch(**es) if isinstance(es, dict)
+                   else es)
+        self.index = index
+        self.mapper = mapper
+        self.criteria = criteria
+        self.doc_type = "run_start"
+        self.doc_properties = {
+            "time": {"type": "date", "format": "epoch_second"},
+            "date": {"type": "date", "format": "strict_date_optional_time"},
+        }
+        self._verified_index = ''
+        return
+
+
+    @classmethod
+    def from_config(cls, config):
+        """
+        Create a new ElasticIndex instance using a configuration dictionary.
+
+        Parameters
+        ----------
+        config : dict
+            The configuration dictionary that describes ElasticIndex.
+            It must contain "databroker-elasticsearch" key.
+
+        Returns
+        -------
+        ElasticIndex
+        """
+        from databroker_elasticsearch.elasticdocument import ElasticDocument
+        cfg = config['databroker-elasticsearch']
+        esdoc = ElasticDocument(cfg['docmap'])
+        rv = cls(es=cfg['host'], index=cfg['index'], mapper=esdoc)
+        return rv
+
+
+    def _generate(self, docs):
+        """Produce transformed Elasticsearch entries that pass the criteria.
+
+        The transformed documents must have an `_id` key which is then used
+        for Elasticsearch unique identifier.
+
+        Parameters
+        ----------
+        docs : iterable
+            The sequence of input documents of dictionary type.
+
+        Yield
+        -----
+        tuple
+            The pairs of (_id, _source) for ES entry identifier and body.
+        """
+        okdocs = (docs if self.criteria is None
+                  else filter(self.criteria, docs))
+        entries = (okdocs if self.mapper is None
+                   else map(self.mapper, okdocs))
+        for e in entries:
+            doc = e.copy()
+            i = doc.pop('_id')
+            yield (i, doc)
+        pass
+
+
+    def _ensure_index_exists(self):
+        """Create and reset the index if it does not exist yet.
+        """
+        if self.index == self._verified_index:
+            return
+        if not self.es.indices.exists(self.index):
+            self.reset()
+            assert self.es.indices.exists(self.index)
+        self._verified_index = self.index
+        # check that short-circuit works or fail with infinite recursion.
+        self._ensure_index_exists()
+        return
+
+
+    def reset(self):
```

Comment:
Should this be a private method? Do we want users to delete all the data in the index?

Reply:
`reset` should be public because it is used from `ElasticCallback.rebuild`
The ElasticIndex class is a convenience interface for ES operations, but users should not work with it directly.